README
Data::Presenter - Reformat database reports

This document refers to version 1.02 of Data::Presenter.  This version was
released December 30, 2005.  If you used versions of Data::Presenter prior to
version 1.00, please read the "Notice of Change of Interface" in the
documentation, as interfaces to several methods have changed and changes need
to be made in the Data::Presenter subclasses from which new() is called.

To install this module on your system, place the tarball archive file in a 
temporary directory and call the following:

% gunzip Data-Presenter-1.02.tar.gz
% tar xf Data-Presenter-1.02.tar
% cd Data-Presenter-1.02
% perl Makefile.PL
% make
% make test
% make install

To better understand the tests as they are run, you may substitute the 
following for the next-to-last-line in the sequence of commands above:

% make test TEST_VERBOSE=1

If you are installing this module over any earlier version, you may 
substitute the following for the last line in the sequence of commands 
above:

% make install UNINST=1

If you are installing this module on a Win32 system with 'nmake', 
substitute 'nmake' for 'make' in the sequence of commands above.

The author has found that trying to install this distribution with 
(a) older (pre-6.16) versions of ExtUtils::MakeMaker, (b) on older 
versions of Perl (e.g., 5.6.0), and (c) on older Linux distributions 
(e.g., RedHat Linux 7.3) can result in a situation where the module's 
Plain Old Documentation, when transformed into manual pages, is not 
automatically placed in the optimal location for reading thru 'man' 
and 'info' (even though it reads perfectly through 'perldoc').  If you 
experience this problem, issue the following at the command line 
(adapted from a suggestion by Michael Schwern on the module-authors 
mailing list):

% perl Makefile.PL INSTALLSITEMAN3DIR=/usr/share/man/man3

Data::Presenter uses the Carp module which is part of the standard Perl 
distribution.  Data::Presenter also uses the List::Compare module by 
the same author.  This module is available from CPAN; versions 0.15 
and later should be used.  Data::Presenter's test suite uses CPAN modules
IO::Capture and IO::Capture::Extended.  These modules are pure Perl and should
install easily with the cpan utility.

For testing purposes, this distribution comes with four sample 
Data::Presenter subclass modules (now located under the t/ directory) 
which illustrate what is required for such a subclass.  In 
addition, the distribution comes with four configuration files, one for 
each sample subclass, which are required to construct a new 
Data::Presenter subclass object.  The distribution comes with four 
sourcefiles, one per subclass.  These sample subclasses are fully 
installed when you install Data::Presenter.  Once you master the 
principles of creating a Data::Presenter invocant subclass you may 
delete these sample files.

Author:  James E. Keenan (jkeenan@cpan.org).
Creation date:  October 25, 2001.  
Last modification date:  December 30, 2005.  
Copyright (c) 2001-2005 James E. Keenan.  United States.  All rights 
reserved.  This is free software and may be distributed under the 
same terms as Perl itself.


Changes
Revision history for Perl extension Data::Presenter

1.02    Thu Dec 29 2005

    - Corrected documentation error in SYNOPSIS:  lack of calling object for
seen_one_column().

    - _validate_args():  Simplified error message:  What was formerly a carp
followed by a croak is now just a croak.  Also, deleted (for now, at
least) reference to the calling method, as this internal subroutine is called
by a variety of methods and subroutines.

    - Corrected bug in Data::Presenter::Combo::Union::_process_secneeded().
Previously, this auxiliary subroutine processed the elements of %{$secneededref}
without sorting them.  This ultimately meant that the data in the non-first
objects passed as arguments to Combo::Union->new() were being assigned into
records within the Combo::Union object in random order.  Since the keys of
that hash are numerical, sorting the keys in numerical order fixes the
problem.  sort_by_column() now DWIMs for Combo::Union objects.  This bug was
not picked up because all the writeformat-type tests in t/04.t were
essentially misspecified.  They also have been corrected.

    - Eliminated commented-out code in Combo::Intersect and Combo::Union.

1.01    Wed Dec 28 2005

    - Deleted erroneous call to 'use Test::DataPresenterSpecial' from
Data::Presenter::Combo::Union.

1.0     Sat Dec 24 2005

    - Version number incremented to 1.0 to alert users to major change of
interface.  writeformat-family of nine methods now takes list of key-value
pairs as arguments rather than simple list (which required keeping track of
order of arguments).  In Data::Presenter::[Package1] subclasses (those from
which objects are created), _init() now has $index assigned directly to
$data{index} rather than being element 0 of an array passed by reference to
$data{index}.  Extensive additional testing.  Use of Devel::Cover for coverage
analysis was used to trim unreachable code and bring all statement, condition
and subroutine coverage up to 100% and to bring all branch coverage except
open() and close() calls up as well.

0.69    Wed Nov 30 2005
    - Partial overhaul including extensive additions to test coverage and
documentation, but as yet no change in functionality.  Because a bug
(incorrect sorting when requesting sorting in descending order) was found,
this version was not uploaded to CPAN.  Instead, it was tarballed so that I
would have a reference point for the next version.  Because of severity of
bug, v0.68 was withdrawn from CPAN.

0.68    Sat Oct 23 2004
	- Corrected substandard code at +1005 Presenter.pm to avoid error "Bizarre copy of array" error when running tests under Devel::Cover

0.67    Sat Oct 23 2004
	- Corrected substandard code at +322 Presenter.pm to avoid error "Bizarre copy of hash in leave" error when running tests under Devel::Cover

0.66    Wed Jun  9 2004
	- POD correction only:  Corrected error in POD for Data::Presenter::Sample Medinsure which resulted in bad POD-to-HTML translation and bad display on search.cpan.org

0.65    Wed Jun  2 2004
	- Corrected Intersect.pm and Union.pm to avoid "Bizarre array assignment" error when installing under Perl 5.8.4 on Darwin.

0.64	Sun Oct  5 2003
	- Added method get_keys_seen();

0.63	Sun Aug 24 2003
	- Added method seen_one_column();

0.62	Sun Apr 13 2003
	- Corrected errors in documentation.

0.61	Sat Apr 12 2003
	- First upload to CPAN.

0.60	Sat Apr 12 2003
	- Begin preparation for public distribution.

Revision history of Data::Presenter prior to public distribution:

Data::Presenter versions 0.3 through 0.39 incorporate the following
corrections and improvements over version 0.2 which was released 10/28/2001.

=over 4

=item *

Introduced C<$index> as one of the variables specified in I<fields.XXX.data>,
imported into the main package and used as metadata inside the
Data::Presenter object.  This eliminates some hard-coding inside subroutines
C<_init()> and C<_extract_rows()> inside Data::Presenter::[subclass] packages. 
It also necessitated a revision of C<_build_sort_formula()>, a large part of
which was extricated into the separate subroutine C<_formula_engine()>.  This
for the first time permitted on index keys which were not entirely numerical.
 So now one can sort on, say, product serial numbers such as C<'24TY-789'>.

=item *

Established a package global hash C<%reserved> within I<Data::Presenter.pm>
and C<%reserved_partial> within I<Data::Presenter::Combo::Intersect.pm> and
I<Data::Presenter::Combo::Union.pm>.  These are hashes of words such as
'fields', 'parameters', 'index' and 'options' which are reserved for current
or future use as keys in the hash blessed into the Data::Presenter object. 
These keys generally have to be excluded when preparing Data::Presenter
selection, sorting and output methods.  The coding for this exclusion is must
easier if one can write:

    unless $reserved{$i} {
    # do something
    }

in contrast to the earlier:

    unless ($i eq 'fields' || $i eq 'parameters') {
    # do something
    }

=item * C<_format_picture_line()> and C<writeHTML()> now format numerical columns
flush-right.

=item *

A bug was fixed in C<_build_sort_formula()> that was causing 'HERNANDEZ' to
precede 'HERNAN' in alphabetical sorts.  This was caused by the internal use
of C<'|'> as the delimiter between array entries.  C<'|'> has a higher ASCII
position than any alphabetical or numerical character.  Hence 'HERNANDEZ|'
has a lower sorting value position than 'HERNAN|'.  This has been corrected
by substituting C<'!'> as the delimiter, since C<'!'> has a lower ASCII value
than any alphabetical or numerical character.  One side effect is that the
character C<'!'> may not appear in data being input into Data::Presenter
objects.

=item *

Clarified error messages in C<_validate_fields()> and C<_analyze_relation()>.

=item *

Up through v0.31, if the operator were to call C<writeformat()>,
C<writeformat_plus_header()> or any combination thereof I<more than once> in a

lib/Data/Presenter/Combo/Union.pm
package Data::Presenter::Combo::Union;
$VERSION = 1.02; # 12-30-2005
@ISA = qw(Data::Presenter::Combo);
use strict;
use warnings;
use Data::Dumper;

our %reserved_partial = (
    'fields'    => 1,
    'index'     => 1,
    'options'   => 1,
);

sub _merge_engine {
    my ($self, $mergeref) = @_;

    my %base                = %{${$mergeref}{base}};
    my %sec                 = %{${$mergeref}{secondary}};
    my %newbase             = %{${$mergeref}{newbase}};
    my %secneeded           = %{${$mergeref}{secfieldsneeded}};
    my %sharedfields        = %{${$mergeref}{sharedfields}};
    
    my @basefields  = @{$base{'fields'}};
    my %seen = ();

    # Build 2 look-up tables showing records found in base and sec
    my %seenbase = _build_lookup_table(\%base);
    my %seensec  = _build_lookup_table(\%sec);

    # Build 3 look-up tables showing whether fields were found in 
    # base only, sec only, or both
    # (At a future date, consider replacing with List::Compare methods)
    my %seenbaseonly = ();
    my %seenseconly = ();
    my %seenboth = ();
    foreach my $rec (keys %seenbase) {
        if (defined $seensec{$rec}) {
            $seenboth{$rec} = 1;
        } else {
            $seenbaseonly{$rec} = 1;
        }
    }
    foreach my $rec (keys %seensec) {
        $seenseconly{$rec} = 1 
            unless ($seenboth{$rec});
    }

    # Work thru the 3 look-up tables to assign values
    my $null = q{};
    foreach my $rec (keys %seenbaseonly) {
        my @values = _process_base(\%base, $rec);
       
        # If an individual record was only found in the base, then, by
        # definition, its 'entries' in the sec are all 'null'.
        # We only need to assign as many nulls as there are keys in
        # %secneeded.
        for (my $p=0; $p < scalar(keys %secneeded); $p++) {
            push(@values, $null);
        }
        $newbase{$rec} = [@values];
    }
    foreach my $rec (keys %seenseconly) {
        my @values;
        for (my $q=0; $q < scalar(@basefields); $q++) {
            my $bf = $basefields[$q];
            if ($sharedfields{$bf}) {
                $values[$q] = $sec{$rec}->[$sharedfields{$bf}[1]];
            } else {
                $values[$q] = $null;
            }
        }
        @values = _process_secneeded(\%secneeded, \%sec, $rec, \@values);
        $newbase{$rec} = [@values];
    }
    foreach my $rec (keys %seenboth) {
        # If a field is seen in both base and sec, we follow a rule that says
        # the base-value for that field gets assigned to the union -- not the
        # sec-value.
        my @values = _process_base(\%base, $rec);
        @values    = _process_secneeded(\%secneeded, \%sec, $rec, \@values);
        $newbase{$rec} = [@values];
    }
    return \%newbase;
}

sub _build_lookup_table {
    my $dataref = shift;
    my %lookup;
    foreach my $rec (keys %{$dataref}) {
        $lookup{$rec} = 1 unless ($reserved_partial{$rec}); # see package lexical above
    }
    return %lookup;
}

sub _process_base {
    my ($baseref, $rec) = @_;
    my @record = @{${$baseref}{$rec}};
    my @values;
    for (my $q=0; $q < scalar(@record); $q++) {
        $values[$q] = $record[$q];
    }
    return @values;
}

sub _process_secneeded {
    my ($secneededref, $secref, $rec, $valuesref) = @_;
    foreach my $r (sort {$a <=> $b} keys %{$secneededref}) {
        my $s = ${$secref}{$rec}->[$r];
        push(@{$valuesref}, $s);
    }
    return @{$valuesref};
}

1;

############################## DOCUMENTATION ##############################

=head1 NAME

Data::Presenter::Combo::Union

=head1 VERSION

This document refers to version 1.02 of Data::Presenter::Combo::Union, released December 30, 2005. 

=head1 DESCRIPTION

This package is a subclass of, and inherits from, Data::Presenter::Combo.  Please see the Data::Presenter documentation to learn how to use Data::Presenter::Combo::Union.

=head1 AUTHOR

James E. Keenan (jkeenan@cpan.org).

Creation date:  October 28, 2001.  Last modification date:  December 30, 2005.
Copyright (c) 2001-5 James E. Keenan.  United States.  All rights reserved.

All data presented in this documentation or in the sample files in the 
archive accompanying this documentation are dummy copy.  The data was 
entirely fabricated by the author for heuristic purposes.  Any resemblance 
to any person, living or dead, is coincidental.

This is free software which you may distribute under the same terms as 
Perl itself.

=cut 

